keps/sig-multicluster/5339-clusterprofile-plugin-credentials/README.md

```diff
@@ -0,0 +1,691 @@
+	// In some environments, the user config may be exactly the same across many clusters
+	// (i.e. call this exec plugin) minus some details that are specific to each cluster
+	// such as the audience.  This field allows the per cluster config to be directly
+	// specified with the cluster info.  Using this field to store secret data is not
+	// recommended as one of the prime benefits of exec plugins is that no secrets need
+	// to be stored directly in the kubeconfig.
+	// +optional
+	Config runtime.RawExtension `json:"config,omitempty"`
+}```
+
+
+#### ClusterProfile Example
+
+Example of a GKE ClusterProfile, which would map to a plugin providing credentials of type `google`:
+```
+apiVersion: multicluster.x-k8s.io/v1alpha1
+kind: ClusterProfile
+metadata:
+ name: my-cluster-1
+spec:
+  displayName: my-cluster-1
+  clusterManager:
+    name: GKE-Fleet
+  credentials:
+    google:
+      cluster:
+        server: https://connectgateway.googleapis.com/v1/projects/123456789/locations/us-central1/gkeMemberships/my-cluster-1
+status:
+  version:
+    kubernetes: 1.28.0
+  properties:
+   - name: clusterset.k8s.io
+     value: some-clusterset
+   - name: location
+     value: us-central1
+```
+
+
+### Configuring plugins in the controller
+
+Plugins are selected by a string which represents the type of credentials that is expected by the cluster, for example, "google" for GKE Clusters.
+This allows the controller to attach a different binary name or path for the binary.
+
+It is expected that the library will have a mapping from its supported type of credentials to the expected binary to call. The library would be fed via a repeated flag `clusterprofile-creds-provider` for ease of use.
+The flag maps a credentials type to the associated binary and potential flags that should be passed. It cannot contain cluster-specific information (which is not known at that time).
+```
+./controller ... --clusterprofile-creds-provider "google='/usr/bin/gke-gcloud-auth-plugin --flag1 value1 --flag2 value2'"
+```
+
+Despite being a flag, we can express the equivalent structure for each Plugin:
+```
+type Provider struct {
+  CredentialsType string
+  ExecutablePath string
+  args []string
+}
+```
+
+Given the plugin is executed directly by the controller, it may expect to have access to the same environment as the controller itself, inclusive of envvars, filesystem and network.
+It is expected that the identity of the plugin is the same as the controller itself.
+
+### Plugin Examples
+
+As an example, we provide pseudocode for plugins that could easily be implemented with the protocol. They are ultrasimplified
+version of the code and structures to convey the idea and not be an implementation example.
+
+#### Secret Reader plugin
+
+This plugin assumes the controller is aware of the list of clusters ahead of time and has created secrets for them.
+It simply reads the token from the secret mapped to the cluster.
+
+```
+func GetToken(clusterName string) string {
```

Comment:
I wonder how we support creating different identities for different applications (i.e.  argoCD vs Kueue) when the input is only clusterName? Where can the library know which app it is generating the token for?

Reply:
yes, valid concern which means this example is not complete enough.
Let me flush out a bit more details. The goal is to show how someone could do it and it's true we shouldn't encourage bad behaviors.

Reply:
flushed a bit.